File: src/filesystem.py
import os
from src.crypto_utils import encrypt_data, decrypt_data

VFS_ROOT = "VFS/home"

def init_fs():
    if not os.path.exists(VFS_ROOT):
        os.makedirs(VFS_ROOT)

def get_real_path(virtual_path):
    clean_path = os.path.normpath(virtual_path).lstrip('\\').lstrip('/')
    real_path = os.path.join(VFS_ROOT, clean_path)
    if not os.path.abspath(real_path).startswith(os.path.abspath(VFS_ROOT)):
        return None
    return real_path

def create_file(path, content, encrypted=True):
    real_path = get_real_path(path)
    if not real_path: return False, "Invalid path"
    os.makedirs(os.path.dirname(real_path), exist_ok=True)
    
    data = content.encode('utf-8')
    if encrypted:
        data = encrypt_data(data)
        
    with open(real_path, "wb") as f:
        f.write(data)
    return True, "File created."

def read_file(path, encrypted=True):
    real_path = get_real_path(path)
    if not real_path or not os.path.exists(real_path):
        return False, "File not found"
        
    with open(real_path, "rb") as f:
        data = f.read()
        
    if encrypted:
        try:
            data = decrypt_data(data)
        except Exception:
            return False, "Decryption failed or file corrupted."
            
    return True, data.decode('utf-8', errors='ignore')

def list_directory(path=""):
    real_path = get_real_path(path)
    if not real_path or not os.path.exists(real_path):
        return []
    items = []
    for item in os.listdir(real_path):
        item_path = os.path.join(real_path, item)
        is_dir = os.path.isdir(item_path)
        items.append({"name": item, "is_dir": is_dir})
    return items

def search_files(query, path=""):
    real_path = get_real_path(path)
    if not real_path or not os.path.exists(real_path):
        return []
    
    results = []
    for root, dirs, files in os.walk(real_path):
        for name in files:
            if query.lower() in name.lower():
                rel = os.path.relpath(os.path.join(root, name), VFS_ROOT)
                results.append(rel)
    return results

def delete_file(path):
    real_path = get_real_path(path)
    if not real_path or not os.path.exists(real_path):
        return False, "File not found"
    try:
        if os.path.isdir(real_path):
            os.rmdir(real_path)
        else:
            os.remove(real_path)
        return True, "Deleted."
    except Exception as e:
        return False, str(e)

File: src/filesystem.cpp
#include "filesystem.h"
#include "crypto_utils.h"
#include <filesystem>
#include <fstream>
#include <iostream>

namespace fs = std::filesystem;

const std::string VFS_ROOT = "VFS_Storage";
const std::string ENCRYPTION_KEY = "secure_os_key_123";

void init_fs() {
    if (!fs::exists(VFS_ROOT)) {
        fs::create_directory(VFS_ROOT);
    }
}

std::vector<std::string> list_directory(const std::string& path) {
    std::vector<std::string> items;
    std::string target = VFS_ROOT;
    if (!path.empty()) target += "/" + path;

    if (fs::exists(target) && fs::is_directory(target)) {
        for (const auto& entry : fs::directory_iterator(target)) {
            items.push_back(entry.path().filename().string());
        }
    }
    return items;
}

std::string read_file(const std::string& path) {
    std::string target = VFS_ROOT + "/" + path;
    if (!fs::exists(target)) return "";

    std::ifstream in(target, std::ios::binary);
    std::string encrypted_content((std::istreambuf_iterator<char>(in)), std::istreambuf_iterator<char>());
    
    return encrypt_decrypt(encrypted_content, ENCRYPTION_KEY);
}

bool create_file(const std::string& path, const std::string& content) {
    std::string target = VFS_ROOT + "/" + path;
    std::ofstream out(target, std::ios::binary);
    if (!out.is_open()) return false;

    std::string encrypted = encrypt_decrypt(content, ENCRYPTION_KEY);
    out.write(encrypted.data(), encrypted.size());
    return true;
}

bool delete_file(const std::string& path) {
    std::string target = VFS_ROOT + "/" + path;
    if (fs::exists(target)) {
        return fs::remove(target);
    }
    return false;
}

std::vector<std::string> search_files(const std::string& query) {
    std::vector<std::string> results;
    for (const auto& entry : fs::recursive_directory_iterator(VFS_ROOT)) {
        if (entry.path().filename().string().find(query) != std::string::npos) {
            results.push_back(entry.path().lexically_relative(VFS_ROOT).string());
        }
    }
    return results;
}
